[Core] Added high-order elements to Kratos core

mnabideltares opened the pull request:
4 elements are added, Line2D4, Line2D5, Triangle2D10, Triangle2D15, Applied on StructuralMechanics

**📝 Added high-order elements to Kratos core**
New high order geometries are added to Kratos core.

Tags: 
- Geometry, element, triangle, line

**🆕 Changelog**
- Added Line2D4N (new file)
- Added Line2D5N (new file)
- Added Triangle2D10N (new file)
- Added Triangle2D15N (new file)
- Added the ghost points locations for the new triangles for gid output
- Implemented new elements in StructuralMechanics (to be extended if needed).
- Added files for unit test for all 4 elements (new files)

mnabideltares committed:
Added high-order elements to Kratos core

4 elements are added, Line2D4, Line2D5, Triangle2D10, Triangle2D15, Applied on StructuralMechanics

mnabideltares committed:
Fix 1: Added high-order elements to Kratos core

4 elements are added, Line2D4, Line2D5, Triangle2D10, Triangle2D15, Applied on StructuralMechanics

loumalouomega commented on applications/StructuralMechanicsApplication/structural_mechanics_application.h:
```diff
@@ -309,6 +309,8 @@ class KRATOS_API(STRUCTURAL_MECHANICS_APPLICATION) KratosStructuralMechanicsAppl
     const SmallDisplacement mSmallDisplacement2D6N;
     const SmallDisplacement mSmallDisplacement2D8N;
     const SmallDisplacement mSmallDisplacement2D9N;
+    const SmallDisplacement mSmallDisplacement2D10N;
```
It would be interesting to extend to TL and UL elements

AlejandroCornejo replied:
I totally agree with @loumalouomega , it could be totally useful for the community

mnabideltares replied:
Thank you @loumalouomega and @AlejandroCornejo 
I already added 4 elemnts:

TotalLagrangianElement2D10N
TotalLagrangianElement2D15N
UpdatedLagrangianElement2D10N
UpdatedLagrangianElement2D15N

Will be in the coming commit (together with some other fixes)

loumalouomega reviewed (commented):
Very nice, it is possible to add a patch test for the new solid elements? (you can see current implementation as reference)

mnabideltares committed:
Fix 2: Added high-order elements to Kratos core

A bug if fixed in test_geometry.cpp

mnabideltares committed:
Fix 3: Added high-order elements to Kratos core

Added elements to StrucMech., fixed issues

rubenzorrilla commented on kratos/geometries/line_2d_4.h:
```diff
@@ -666,15 +666,15 @@ class Line2D4 : public Geometry<TPointType>
     SizeType EdgesNumber() const override
     {
-        return 2;
+        return 3;
```
I think this should be 2.

mnabideltares replied:
Thank you @rubenzorrilla 
What I understand, the line elements have always 2 edges?
How the faces are defined? Is the FacesNumber=0 for all line elements?

rubenzorrilla replied:
Well, I think there is no formal definition. If I'm not remembering wrong, in the @KratosMultiphysics/technical-committee we agreed to consider as edges/faces both endpoints of the line.

rubenzorrilla commented on kratos/geometries/line_2d_5.h:
```diff
@@ -670,15 +670,15 @@ namespace Kratos
         SizeType EdgesNumber() const override
         {
-            return 2;
+            return 4;
```
This one as well.

mnabideltares committed:
Merge branch 'master' into 10083-migrate-the-new-designed-elements-from-geomechanics-branch-to-kratos-application

mnabideltares committed:
Fix 3: Added high-order elements to Kratos

mnabideltares committed:
Fix 4: Added high-order elements to Kratos

loumalouomega commented on kratos/geometries/geometry.h:
```diff
@@ -1221,7 +1221,8 @@ class Geometry
         } else if (LumpingMethod == LumpingMethods::DIAGONAL_SCALING) {
             IntegrationMethod integration_method = GetDefaultIntegrationMethod();
-            integration_method = static_cast<IntegrationMethod>(static_cast<int>(integration_method) + 1);
+            int j = std::min(static_cast<int>(integration_method) + 1, 4);
```
```suggestion
            const int j = std::min(static_cast<int>(integration_method) + 1, 4);
```
Only because of this (modify geometry.h) we will need the approval from @KratosMultiphysics/technical-committee

mnabideltares replied:
I used this as ad-hoc fix. I am using Gauss_5 for Triangle2D15. In that part (in geometry.h), in order to get higher accuracy, it adds one order more to Gauss, and tries to make it Gauss_6 (in order to calculta N^2). However, this option is not available in the code.
Another remedy for this is to extent Gauss points to higher orders. Do you think it will be a more suitable option?

loumalouomega commented on kratos/includes/gid_io.h:
```diff
@@ -265,6 +265,12 @@ class KRATOS_API(KRATOS_CORE) GidIO
         mGidMeshContainers.push_back( TMeshContainer(
                                           GeometryData::KratosGeometryType::Kratos_Triangle2D6,
                                           GiD_Triangle, "Kratos_Triangle2D6_Mesh" ) );
+        mGidMeshContainers.push_back(TMeshContainer(
```
I don't know if GiD supports these geometries. @jginternational ?

mnabideltares replied:
Actually GiD can't read these geometries. However, I added this part for consistency with the other geometries.
Do I replace it with an error message?

jginternational replied:
GiD allows triangles with 3 and 6 nodes. There are no plans to add more triangular geometries

loumalouomega replied:
In that case, I would remove, or we will have not functional GiD outputs. An alternative is to "patch" the new geometries into old geometries. So for example, divide it into linear triangles.

jginternational replied:
@loumalouomega @mnabideltares 
We have this list of valid elements in GiD and their connectivities: [https://gidsimulation.atlassian.net/wiki/spaces/GRM/pages/2385547733/Element+type](https://gidsimulation.atlassian.net/wiki/spaces/GRM/pages/2385547733/Element+type)

rubenzorrilla replied:
For the moment I'd remove as @loumalouomega suggests. In the future we can (in an independent PR) think about patching the visualization of such high order geometries by using low order ones.

mnabideltares replied:
Removed.
I also think it has to be done in an independent PR as @loumalouomega  and @rubenzorrilla  suggested.

loumalouomega commented on applications/StructuralMechanicsApplication/structural_mechanics_application.h:
```diff
@@ -420,11 +426,15 @@ class KRATOS_API(STRUCTURAL_MECHANICS_APPLICATION) KratosStructuralMechanicsAppl
 
     const SmallDisplacementLineLoadCondition<2> mSmallDisplacementLineLoadCondition2D2N;
     const SmallDisplacementLineLoadCondition<2> mSmallDisplacementLineLoadCondition2D3N;
+    const SmallDisplacementLineLoadCondition<2> mSmallDisplacementLineLoadCondition2D4N;
```
It would be interesting adding some patch tests for the corresponding elements. See the tests in the structural mechanics as example

mnabideltares replied:
in the form of cpp as unit-test? Or real separate test cases?

loumalouomega replied:
I would extend the current patch_tests in the tests folder

mnabideltares replied:
Added patch tests

loumalouomega reviewed (changes requested):
Comments

mnabideltares committed:
Fix 5: Added high-order elements to Kratos

Removed the GID output for the high order elements

loumalouomega reviewed (changes requested):
The only thing missing is extending the patch_tests from the tests folder in StructuralMechanicsApplication

mnabideltares committed:
Added patch tests for elements of 10 and 15 nodes, for TL and UL

mnabideltares committed:
TestLengthLine2D5 is updated

mnabideltares committed:
Added the number of gauss points for Triangle2D10N

Gauss_3 to Gauss_4: leads to more accurate results based on patch tests for TL and UL

mnabideltares committed:
lumping factors of DIAGONAL_SCALING for Triangle2D10N are updated

loumalouomega commented on kratos/tests/cpp_tests/geometries/test_triangle_2d_10.cpp:
```diff
@@ -0,0 +1,216 @@
+//                   Multi-Physics
+//
+//  License:		 BSD License
+//					 Kratos default license: kratos/license.txt
```
```suggestion
//  License:               BSD License
//                              Kratos default license: kratos/license.txt
```

loumalouomega commented on kratos/tests/cpp_tests/geometries/test_triangle_2d_15.cpp:
```diff
@@ -0,0 +1,243 @@
+//                   Multi-Physics
+//
+//  License:		 BSD License
+//					 Kratos default license: kratos/license.txt
```
```suggestion
//  License:               BSD License
//                              Kratos default license: kratos/license.txt
```

loumalouomega commented on kratos/geometries/triangle_2d_15.h:
```diff
@@ -0,0 +1,1609 @@
+//                   Multi-Physics
+//
+//  License:		 BSD License
+//					 Kratos default license: kratos/license.txt
```
```suggestion
//  License:               BSD License
//                              Kratos default license: kratos/license.txt
```

loumalouomega commented on kratos/geometries/triangle_2d_10.h:
```diff
@@ -0,0 +1,1356 @@
+//                   Multi-Physics
+//
+//  License:		 BSD License
+//					 Kratos default license: kratos/license.txt
```
```suggestion
//  License:               BSD License
//                              Kratos default license: kratos/license.txt
```

loumalouomega commented on kratos/geometries/line_2d_5.h:
```diff
@@ -0,0 +1,1171 @@
+//                   Multi-Physics
+//
+//  License:		 BSD License
+//					 Kratos default license: kratos/license.txt
```
```suggestion
//  License:               BSD License
//                              Kratos default license: kratos/license.txt
```

loumalouomega commented on kratos/geometries/line_2d_4.h:
```diff
@@ -0,0 +1,1143 @@
+//                   Multi-Physics
+//
+//  License:		 BSD License
+//					 Kratos default license: kratos/license.txt
```
```suggestion
//  License:               BSD License
//                              Kratos default license: kratos/license.txt
```

loumalouomega committed:
Update headers

loumalouomega committed:
Pragma once
